Firmware/communication/can_simple.cpp

```diff
@@ -0,0 +1,423 @@
+            case MSG_GET_ENCODER_ERROR:
+                get_encoder_error_callback(axis, msg);
+                break;
+            case MSG_GET_SENSORLESS_ERROR:
+                get_sensorless_error_callback(axis, msg);
+                break;
+            case MSG_SET_AXIS_NODE_ID:
+                set_axis_nodeid_callback(axis, msg);
+                break;
+            case MSG_SET_AXIS_REQUESTED_STATE:
+                set_axis_requested_state_callback(axis, msg);
+                break;
+            case MSG_SET_AXIS_STARTUP_CONFIG:
+                set_axis_startup_config_callback(axis, msg);
+                break;
+            case MSG_GET_ENCODER_ESTIMATES:
+                get_encoder_estimates_callback(axis, msg);
+                break;
+            case MSG_GET_ENCODER_COUNT:
+                get_encoder_count_callback(axis, msg);
+                break;
+            case MSG_MOVE_TO_POS:
+                move_to_pos_callback(axis, msg);
+                break;
+            case MSG_SET_POS_SETPOINT:
+                set_pos_setpoint_callback(axis, msg);
+                break;
+            case MSG_SET_VEL_SETPOINT:
+                set_vel_setpoint_callback(axis, msg);
+                break;
+            case MSG_SET_CUR_SETPOINT:
+                set_current_setpoint_callback(axis, msg);
+                break;
+            case MSG_SET_VEL_LIMIT:
+                set_vel_limit_callback(axis, msg);
+                break;
+            case MSG_START_ANTICOGGING:
+                start_anticogging_callback(axis, msg);
+                break;
+            case MSG_SET_TRAJ_A_PER_CSS:
+                set_traj_A_per_css_callback(axis, msg);
+                break;
+            case MSG_SET_TRAJ_ACCEL_LIMITS:
+                set_traj_accel_limits_callback(axis, msg);
+                break;
+            case MSG_SET_TRAJ_VEL_LIMIT:
+                set_traj_vel_limit_callback(axis, msg);
+                break;
+            case MSG_GET_IQ:
+                get_iq_callback(axis, msg);
+                break;
+            case MSG_GET_SENSORLESS_ESTIMATES:
+                get_sensorless_estimates_callback(axis, msg);
+                break;
+            case MSG_RESET_ODRIVE:
+                NVIC_SystemReset();
+                break;
+            case MSG_GET_VBUS_VOLTAGE:
+                get_vbus_voltage_callback(axis, msg);
+                break;
+            default:
+                break;
+        }
+    }
+}
+
+void CANSimple::nmt_callback(Axis* axis, CAN_message_t& msg) {
+    // Not implemented
+}
+
+void CANSimple::estop_callback(Axis* axis, CAN_message_t& msg) {
+    axis->error_ |= Axis::ERROR_ESTOP_REQUESTED;
+}
+
+void CANSimple::get_motor_error_callback(Axis* axis, CAN_message_t& msg) {
+    if (msg.rtr) {
+        CAN_message_t txmsg;
+        txmsg.id = axis->config_.can_node_id << NUM_CMD_ID_BITS;
+        txmsg.id += MSG_GET_MOTOR_ERROR;  // heartbeat ID
+        txmsg.isExt = false;
+        txmsg.len = 8;
```

Comment:
This should probably be `txmsg.len = 4`.

Or maybe even `if (txmsg.len != 4) return;`. Based on [this explanation](http://www.esd-electronics-usa.com/CAN-Remote-Frames.html) collisions can occur if remote frame and data frame use different length fields so it may be good to discourage this by not responding, even if it sometimes works.

Reply:
I don't understand what you are referring to.

Reply:
We can use 4 bytes in this message if we want, but right now all messages are 8 bytes.  Common thing to do for whatever reason.

But why would I check for txmsg.len == 4 when I'm the one who is deciding how long the message is?

Reply:
Ah ok didn't know that was common.

The problem with remote frames is that if two nodes send the same remote frame but with a different length field, they can both win arbitration and then collide during the data field (I worded this a bit wrong before).

So it makes somewhat sense to encourage all client implementations to send the same data length field, by just ignoring any non-compliant remote frame. I mean once you received a non-compliant remote frame the collision was already evaded so you might as well just serve it, but in the long run the "ignore strategy" will kill off non-compliant implementations.